Add Resource trait; rewrite client stack around it

anguslees opened the pull request:
Major changes to the layering all over.  We now have:

- Abstract HttpService / HttpUpgradeService trait for a basic HTTP request
  ... and an implementation of that using hyper

- Abstract ApiService trait that represents the Kubernetes API
  ... and an implementation of that using HttpService

- Parameterised ResourceClient that captures the end-to-end
  request+response for an arbitrary Resource

anguslees committed:
Add Resource trait; rewrite client stack around it

Major changes to the layering all over. `holding` is dead. We now have:

- Abstract HttpService / HttpUpgradeService trait for a basic HTTP request
  ... and an implementation of that using hyper

- Abstract ApiService trait that represents the Kubernetes API
  ... and an implementation of that using HttpService

- Parameterised ResourceClient that captures the end-to-end
  request+response for an arbitrary Resource

anguslees commented:
I acknowledge that this is a big code dump, and effectively unreviewable (except as new code).  I can attempt to slice it up into self-contained dependency layers if you would prefer, although this is hard when working as diffs from the previous code.

Mostly I want to get this merged, so I can do the much-needed 2018 edition conversion, and then jump on the new async-await bandwagon (as it gets stabilised).

rbtcollins commented:
If waiting till the weekend is ok, I can commit to reading it in detail then; I've been sick last week and need to actually focus on work hard this week (and am not yet fully well).

If waiting is no ok, please merge it and ce la vie.

anguslees committed:
Add dynamic client example

anguslees committed:
Remove main() wrapper in examples

Simplify examples by removing the wrapper around `main()`.  The new
default `main() -> Result` printer is good enough.

rbtcollins commented on README.md:
Perhaps we should call that hyperclient or something

anguslees replied:
Hyper is meant to be an implementation detail and I wouldn't expect there to ever be a family of other clients (unless another http client becomes the new shiny, in which case we would probably transition in place rather than start all over again)

rbtcollins replied:
Does the type leak /at all/ ? If it leaks (in dependency pins, in signatures, but perhaps we can exclude dll/soname references), then I think we set ourselves up for pain saying we'd transition in-place :). Perhaps I misjudge. Happy for us to decide later.

anguslees replied:
I leak the standard "http" crate and I think I will need/want to leak some tokio references in the near future.

rbtcollins replied:
well, lets iterate.

rbtcollins commented on README.md:
```diff
@@ -32,30 +32,27 @@ Example of listing all the pods in `kube-system` namespace.
+use kubernetes_api::core::v1::Pods;
+use kubernetes_apimachinery::meta::NamespaceScope;
+use kubernetes_apimachinery::meta::v1::ListOptions;
+use kubernetes_client::{Client,ListOptions};
```
I worry about grabbing so many top level names; are they all free?  kubernetes we already have right?

anguslees replied:
Agreed.  I can roll everything into a single crate if we prefer. I think the only one I might want separate is the "api" crate, since this needs a release for each upstream k8s release and i think it will be easier to manage that on a separate lifecycle.

rbtcollins replied:
If the names are free I have no objection to the sprawl. lets see ... seems ok. We can always consolidate later.

rbtcollins commented on api/src/policy/v1beta1.rs:
```diff
@@ -0,0 +1,41 @@
+    fn kind() -> &'static str {
+        "Eviction"
+    }
+}
```
I'm thinking we want a derive macro for these.

anguslees replied:
I have that macro already, I'm waiting for this PR to be finalised before publishing it.

rbtcollins commented on api/src/core/v1.rs:
```diff
@@ -13,34 +23,259 @@ pub const GROUP_VERSION: GroupVersion = GroupVersion {
+// TODO: The stream errors should probably be some other error type.
+// Also: StreamItem should probably be a Buf, or replace the whole
+// thing with AsyncRead/Write.
+pub type ByteStream = Box<Stream<Item = Vec<u8>, Error = Error> + Send>;
```
It would be nice to show this as unstable someway

anguslees replied:
I don't think we have a way to do this (I don't think std's `#[unstable]` annotation means anything in user crates) - apart from what the crate-level "0.1" semver communicates (there is no API stability at present).

rbtcollins commented on apimachinery/src/meta/mod.rs:
```diff
@@ -268,3 +286,334 @@ fn gr_fromstr() {
+/// Zero-sized struct that serializes to/from apiVersion/kind struct
+/// based on type parameter.
+#[derive(Default, Debug, Clone)]
+pub struct TypeMetaImpl<T>(PhantomData<T>);
```
very cute, me liky

rbtcollins commented on apimachinery/src/meta/mod.rs:
```diff
@@ -268,3 +286,334 @@ fn gr_fromstr() {
+    }
+}
+
+pub trait ResourceScope {
```
This feels a little awkward, particularly with the impls below, I guess we'll see where its used later...

anguslees replied:
Yes, I'm not completely happy with these "name" bits. Alternative suggestions welcome :-\

rbtcollins commented on apimachinery/src/request.rs:
```diff
@@ -0,0 +1,187 @@
+}
+
+#[derive(Debug)]
+pub struct Request<B, O> {
```
I think it would be useful to say BODY or Body rather than B here, similar OPTS - we're not running out of space :P

rbtcollins commented on apimachinery/src/client.rs:
```diff
@@ -0,0 +1,664 @@
+    ) -> impl Future<Item = R::Item, Error = Error> + Send {
+        let gvr = self.resource.gvr();
+
+        let req = Request {
```
The high repetition here makes me think a Request constructor from gvr + name + subresource + opts + contentype would help.

anguslees replied:
Agreed 100%.  I was trying (unsuccessfully) to halt the rampant scope creep in this PR.  I will retro-fit a builder onto these in a followup.

rbtcollins reviewed (approved):
I have a number of suggestions for improvement but nothing that needs to prevent this being merged. Sorry for the delay... Strep throat :(
